From 4c20d29c296d6bce6a0e173a26011d8c406aba6d Mon Sep 17 00:00:00 2001
From: Willy Tarreau <w@1wt.eu>
Date: Sat, 14 Jun 2014 11:41:36 +0200
Subject: [PATCH 24/30] BUG/MINOR: connection: make proxy protocol v1 support
 the UNKNOWN protocol

If haproxy receives a connection over a unix socket and forwards it to
another haproxy instance using proxy protocol v1, it sends an UNKNOWN
protocol, which is rejected by the other side. Make the receiver accept
the UNKNOWN protocol as per the spec, and only use the local connection's
address for this.
---
 src/connection.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/src/connection.c b/src/connection.c
index 91ad0b6..93db7e5 100644
--- a/src/connection.c
+++ b/src/connection.c
@@ -286,7 +286,7 @@ int conn_recv_proxy(struct connection *conn, int flag)
 	}
 
 	line += 6;
-	if (trash.len < 18) /* shortest possible line */
+	if (trash.len < 9) /* shortest possible line */
 		goto missing;
 
 	if (!memcmp(line, "TCP4 ", 5) != 0) {
@@ -391,8 +391,12 @@ int conn_recv_proxy(struct connection *conn, int flag)
 		((struct sockaddr_in6 *)&conn->addr.to)->sin6_port          = htons(dport);
 		conn->flags |= CO_FL_ADDR_FROM_SET | CO_FL_ADDR_TO_SET;
 	}
+	else if (memcmp(line, "UNKNOWN\r\n", 9) == 0) {
+		/* This can be a UNIX socket forwarded by an haproxy upstream */
+		line += 9;
+	}
 	else {
-		/* The protocol does not match something known (TCP4/TCP6) */
+		/* The protocol does not match something known (TCP4/TCP6/UNKNOWN) */
 		conn->err_code = CO_ER_PRX_BAD_PROTO;
 		goto fail;
 	}
-- 
1.8.3.2

